### PR TITLE
[SYCL] Add section about buffer limitations

### PR DESCRIPTION
**Pull request by Bensuo**
- Adds section to the spec on buffer limitations around the copy-back mechanism.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -1057,6 +1057,25 @@ extended either for the lifetime of the graph (including both modifiable graphs
...
 required by the graph (such as after being replaced through executable graph update).
 
+=== Buffer Limtations
+
+Because of the delayed execution of a recorded graph it is not possible to support
+code which relies on the copy-back on destruction behaviour of buffers.
+Typically applications would rely on this behaviour to do work on the host which
+cannot inherently be captured inside a command graph. Thus when recording to a graph
+it is an error to submit a command which has an accessor on a buffer which would
+cause a write-back to happen. Using an incompatible buffer in this will result
```
Comment (EwanC) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]: in this **case**

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 2]
```diff
@@ -1057,6 +1057,25 @@ extended either for the lifetime of the graph (including both modifiable graphs
 and the executable graphs created from them) or until the buffer is no longer
 required by the graph (such as after being replaced through executable graph update).
 
+=== Buffer Limtations
+
+Because of the delayed execution of a recorded graph it is not possible to support
+code which relies on the copy-back on destruction behaviour of buffers.
+Typically applications would rely on this behaviour to do work on the host which
+cannot inherently be captured inside a command graph. Thus when recording to a graph
```
Comment (EwanC) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc:1069]: "recording to a graph" - would it also be an error to create a graph node using the explicit API which uses an accessor that causes write-back? we should say so one way or another

Comment (Bensuo) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc:1069]: After some thought I think it is not an error, so I have added a note that the restrictions do not apply in the explicit API. However, I have noted that use of the copy-back mechanism should still be avoided in user code.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc (hunk 1 above)
Comment (Bensuo) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]: Fixed thanks!

**Review by EwanC — APPROVED** [review]

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 3]
```diff
@@ -1057,6 +1057,30 @@ extended either for the lifetime of the graph (including both modifiable graphs
...
+in a synchronous error being thrown with error code `invalid`.
+
+The copy-back mechanism can be disabled explicitly for buffers with attached host
+storage using either `buffer::set_final_data(nullptr)` or
+`buffer::set_copy_back(false)`.
+
+When using the explicit graph API this restriction on buffers does not apply and
+the above error will not be thrown. However, because of the extension of storage
+lifetimes users should avoid relying on the copy-back mechanism in code using the
+explicit API.
```
Comment (reble) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]: Would it make sense to start the section with this paragraph? Like: The following restrictions do not apply to the explicit API ... 

Advantage: if you're only interested in the explicit API you could skip the whole section.

**Review by reble — APPROVED** [review]

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc (hunk 3 above)
Comment (Bensuo) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]: Yes I think that makes sense to me! I've moved the general warning against using copy-back to the above section on storage lifetimes and changed the section title to reflect that this applies only to the record and replay API.

**Pull request merged**